src/lab15.py

```diff
@@ -1234,25 +1234,23 @@ def load(self, url, body=None):
             self.rules.extend(CSSParser(body.decode("utf8")).parse())
 
         images = [node
-                   for node in tree_to_list(self.nodes, [])
-                   if isinstance(node, Element)
-                   and node.tag == "img"
-                   and "src" in node.attributes]
+            for node in tree_to_list(self.nodes, [])
+            if isinstance(node, Element)
+            and node.tag == "img"
+            and node.attributes.get("src")]
```

Comment:
You should show the broken image icon even if the `src` attribute is omitted

Reply:
I did some tests in a Chrome variant and it looked like you needed a nonempty src. Just to be sure, I just looked it up in the spec:

https://html.spec.whatwg.org/multipage/embedded-content.html#the-img-element

Way down in this section it tells what the image presents as:

> If the [src](https://html.spec.whatwg.org/multipage/embedded-content.html#attr-img-src) attribute is not set and either the [alt](https://html.spec.whatwg.org/multipage/embedded-content.html#attr-img-alt) attribute is set to the empty string or the [alt](https://html.spec.whatwg.org/multipage/embedded-content.html#attr-img-alt) attribute is not set at all
> The element [represents](https://html.spec.whatwg.org/multipage/dom.html#represents) nothing.

> Otherwise
> The element [represents](https://html.spec.whatwg.org/multipage/dom.html#represents) the text given by the [alt](https://html.spec.whatwg.org/multipage/embedded-content.html#attr-img-alt) attribute.

which points out an interesting thing, that an Img element without src but with alt is actually a kind of text node… I don’t feel like implementing this but it could be a footnote.

another thing not totally clear in the spec is what it means by a the src attribute not being set. In my tests, an empty, set src attribute still doesn’t trigger fallback behavior, and text earlier in the spec suggests you should strip whitespace, but it’s not obvious to me. I haven’t read this part of the spec in full so I might be missing something.

Reply:
The part of the spec you mentioned is in an exercise already. But `data:text/html,<img alt=foo>` does show the broken image in Chrome.

Reply:
Ok, this definitely isn't too important but we should understand for ourselves. Here's a series of tests I did, all with a doctype so we're not in quirks mode:

[ ] `<img>`
[ ] `<img src>`
[ ] `<img src=" ">`
[x] `<img src="a">`
[ ] `<img alt>`
[ ] `<img src alt>`
[ ] `<img src=" " alt>`
[ ] `<img src="a" alt>`
[x] `<img alt=" ">`
[x] `<img src="a" alt=" ">`

Reply:
The weirdest one is `<img src="a" alt>`, which seems like it should show a placeholder but doesn't. I can't explain that. If you ignore that, then a nonempty `src` or a nonempty `alt` is enough, but perhaps we should leave the non-empty `alt` case as an exercise?

Reply:
I think it's because a bare `alt` isn't nonempty.

Reply:
Ok, let's discuss this tomorrow.